### ISSUE TITLE
people counting img classification: Drop out/BN/SGD

### ISSUE THREAD

**Issue by songhee-cho** [SongheeCho/FMCW-radar-img-classification#1]
### **학습 목적**🍐

* **특정 공간 내에 존재하는 사람의 숫자(0, 1, 2, 3)를 기준으로 image data를 data0, data1, data2, data3으로 classification**
   * FMCW radar를 이용하여 특정 공간 내부에 존재하는 사람의 수(0~3명)를 알 수 있는 image data 획득
   * 사람 수가 N명일 때 획득된 image data를 dataN로 정의(N = 0, 1, 2, 3)

### **model의 layer 구성**🧙🏻

model의 layer를 도식화한 표를 첨부해주세요.

### **model을 구현하기 위한 코드**✍🏻

* **_tensorflow ver 2.0 사용 학습 환경 구성_**
* **_training/validation/test dataset 확보_**
```python
# import tensorflow and other methods

import tensorflow as tf

from tensorflow.keras import datasets, layers, models
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Conv2D, Activation, Flatten, Dropout, MaxPooling2D, BatchNormalization
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from scipy import io
from tensorflow.keras.utils import plot_model

import os
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'
import numpy as np
import matplotlib.pyplot as plt

# load data
# capon beam forming한 결과 image
data_mat_file = io.loadmat('people0to3.mat')
train_label_mat_file = io.loadmat('people0to3_train_label_new.mat')
validation_label_mat_file = io.loadmat('people0to3_validation_label_new.mat')
test_label_mat_file = io.loadmat('people0to3_test_label_new.mat')

data0 = data_mat_file['data0']
data1 = data_mat_file['data1']
data2 = data_mat_file['data2']
data3 = data_mat_file['data3']

# train data, validation data, test data 설정
train_data0 = data0[0:800,:,:]
train_data1 = data1[0:800,:,:]
train_data2 = data2[0:800,:,:]
train_data3 = data3[0:800,:,:]

validation_data0 = data0[800:900,:,:]
validation_data1 = data1[800:900,:,:]
validation_data2 = data2[800:900,:,:]
validation_data3 = data3[800:900,:,:]

test_data0 = data0[900:1000,:,:]
test_data1 = data1[900:1000:,:,:]
test_data2 = data2[900:1000,:,:]
test_data3 = data3[900:1000,:,:]

train_data = np.vstack([train_data0, train_data1, train_data2, train_data3])
train_labels_new = train_label_mat_file['train_labels_new']

validation_data = np.vstack([validation_data0, validation_data1, validation_data2, validation_data3])
validation_labels_new = validation_label_mat_file['validation_labels_new']

test_data = np.vstack([test_data0, test_data1, test_data2, test_data3])
test_labels_new = test_label_mat_file['test_labels_new']
```

* **_data 전처리 및 normalization_**
```python
# data preparation, normalization
data0 = data0/255.0
data1 = data0/255.0
data2 = data0/255.0
data3 = data0/255.0
img_height = 120
img_width = 128
train_data = train_data[:,:,:, np.newaxis].astype('float32')
validation_data = validation_data[:,:,:, np.newaxis].astype('float32')
test_data = test_data[:,:,:, np.newaxis].astype('float32')
```

* **_CNN model 구현_**
```python
# create the model
# filters = 16 use 16 kernels
# use 3*3 sized kernel
model = Sequential()
# 1st 2D Conv layer
# 16 filters
# padding=same: size 유지, same: 필터의 사이즈가 k이면 사방으로 k/2
conv1=Conv2D(filters=16, kernel_size=(5, 5), strides=(1, 1), padding='same', input_shape=(img_height, img_width, 1))
model.add(conv1)
# 1st batch normalization
bns1 = BatchNormalization()
model.add(bns1)
# activation relu
act1 = Activation("relu")
model.add(act1)
# 1st max pooling(down sampling)
maxPool1 = MaxPooling2D(pool_size=2, strides=None, padding='valid')
model.add(maxPool1)

# 2nd 2D Conv layer
# 32 filters
conv2=Conv2D(filters=32, kernel_size=(5, 5), strides=(1, 1), padding='same')
model.add(conv2)
# 2nd batch normalization
bns2 = BatchNormalization()
model.add(bns2)
# activation relu
act2 = Activation("relu")
model.add(act2)


# 3rd 2D Conv layer
# 64 filters
conv3 = Conv2D(filters=64, kernel_size=(3, 3), strides=(1, 1), padding='same')
model.add(conv3)
# 3rd batch normalization
bns3 = BatchNormalization()
model.add(bns3)
# activation relu
act3 = Activation("relu")
model.add(act3)

# flatten
flttn1 = Flatten()
model.add(flttn1)
# hidden fully connected layer
dns1 = Dense(64)
model.add(dns1)
# activation relu
act4 = Activation("relu")
model.add(act4)
# dropout
#drp1 = Dropout(keep_prob)
#model.add(drp1)
# 출력 fully connected layer
dns2 = Dense(num_classes)
model.add(dns2)
# activation relu
act5 = Activation("softmax")
model.add(act5)
```
* **_model 확인_**
```python
# model summary
model.summary()
```
* **_model 학습 결과 확인_**
```python
# compile model
model.compile(optimizer='sgd', loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])
# train the model
history = model.fit(train_data, train_labels_new,
                    validation_data=(validation_data, validation_labels_new),
                    epochs=50, batch_size=64, verbose=2)
epochs = 50
acc = history.history['accuracy']
val_acc = history.history['val_accuracy']

loss = history.history['loss']

val_loss = history.history['val_loss']
epochs_range = range(epochs)
```
* **_model 확인_**
```python
# check accuracy and loss
plt.figure(figsize=(12, 8))
plt.subplot(1, 2, 1)
plt.plot(epochs_range, acc, label='Training Accuracy')
plt.plot(epochs_range, val_acc, label='Validation Accuracy')
plt.legend(loc='lower right')
plt.title('Training and Validation Accuracy')

plt.subplot(1, 2, 2)
plt.plot(epochs_range, loss, label='Training Loss')
plt.plot(epochs_range, val_loss, label='Validation Loss')
plt.legend(loc='upper right')
plt.title('Training and Validation Loss')
plt.show()

keep_prob = 1
score = model.evaluate(test_data, test_labels_new, verbose=2)
print('Test loss:', score[0])
print('Test accuracy:', score[1])
```
* **_학습 model로 predict한 결과 저장_**
```python
# cnn model로 test data를 추정한 결과값을 저장
test_data = np.vstack([test_data0, test_data1, test_data2, test_data3])
test_data = test_data[:,:,:, np.newaxis].astype('float32')
predicted_result = model.predict(test_data)
```

* **_잘못 분류된 case 나열_**
```python
# wrong example check
# cnn model로 test data를 추정한 결과값을 저장
import random
#/255 중복으로 normalization 여러번 하면 오류 생김
predicted_labels = np.argmax(predicted_result, axis=1)

wrong_result = []

for n in range(0, (len(test_labels_new))):
    if predicted_labels[n] != test_labels_new[n]:
        wrong_result.append(n)

samples = random.choices(population=wrong_result, k=16)

count = 0
nrows = ncols = 4

plt.figure(figsize=(12,8))

for n in samples:
    count += 1
    plt.subplot(nrows, ncols, count)
    plt.imshow(test_data[n].reshape(120, 128))
    tmp = "Label:" + str(test_labels_new[n]) + ", Prediction:" + str(predicted_labels[n])
    plt.title(tmp)

plt.tight_layout()
plt.show()
print("wrongly classified cases: ")
print(len(wrong_result))
```

* **_제대로 분류된 case 나열_**
```python
right_result = []

for n in range(0, len(test_labels_new)):
    if predicted_labels[n] == test_labels_new[n]:
        right_result.append(n)

samples = random.choices(population=right_result, k=16)

count = 0
nrows = ncols = 4

plt.figure(figsize=(12,8))

for n in samples:
    count += 1
    plt.subplot(nrows, ncols, count)
    plt.imshow(test_data[n].reshape(120, 128))
    tmp = "Label:" + str(test_labels_new[n]) + ", Prediction:" + str(predicted_labels[n])
    plt.title(tmp)

plt.tight_layout()
plt.show()
print("well classified cases: ")
print(len(right_result))
```

* **_feature map 확인_**
```python
# feature map
first_layer = model.layers[0]
print(first_layer)
print(model.input)

first_activation = models.Model(inputs=model.input, outputs=first_layer.output)
activation = first_activation.predict(test_data[:400])

print(activation.shape)

fig = plt.figure(figsize=(12,8))
for i in range(16):
    fig.add_subplot(4, 4, i+1)
    plt.imshow(activation[0, :, :, i])

fig = plt.figure(figsize=(12,8))
for i in range(16):
    fig.add_subplot(4, 4, i+1)
    plt.imshow(activation[100, :, :, i])

fig = plt.figure(figsize=(12,8))
for i in range(16):
    fig.add_subplot(4, 4, i+1)
    plt.imshow(activation[200, :, :, i])

fig = plt.figure(figsize=(12,8))
for i in range(16):
    fig.add_subplot(4, 4, i+1)
    plt.imshow(activation[300, :, :, i])
```

* **_사용한 Conv filter 확인_**
```python
# filter
fig = plt.figure(figsize=(12,8))
for i in range(16):
    fig.add_subplot(4, 4, i+1)
    plt.imshow(first_layer.kernel[:, :, 0, i])
```

* **_Grad-CAM for data0_**
```python
test_data = np.vstack([test_data0, test_data1, test_data2, test_data3])
test_data = test_data[:,:,:, np.newaxis]
print(test_data.shape)

DATA0_CLASS_INDEX = 10
test_data_exmpl_0 = test_data[DATA0_CLASS_INDEX,:,:]
test_data_exmpl_0 = np.expand_dims(test_data_exmpl_0, axis=0)
print(test_data_exmpl_0.shape)

import cv2

conv_layer = model.layers[8]
img_tensor_0 = test_data_exmpl_0
heatmap_model = tf.keras.models.Model([model.inputs], [conv_layer.output, model.output])

# Get the score for target class
with tf.GradientTape() as tape:
    conv_outputs_0, predictions_0 = heatmap_model(img_tensor_0)
    loss_0 = predictions_0[:, np.argmax(predictions_0[0])]

# Extract filters and gradients
output_0 = conv_outputs_0[0]
grads_0 = tape.gradient(loss_0, conv_outputs_0)[0]

# Average gradients spatially
weights_0 = tf.reduce_mean(grads_0, axis=(0, 1))
print(weights_0.shape)

# Build a ponderated map of filters according to gradients importance
cam_0 = np.ones(output_0.shape[0:2], dtype=np.float32)
print(cam_0.shape)

for index, w in enumerate(weights_0):
    cam_0 += w * output_0[:, :, index]

# Heatmap visualization
cam_0 = cv2.resize(cam_0.numpy(), (128, 120))
cam_0 = np.maximum(cam_0, 0)
heatmap_0 = cam_0
heatmap_0 /= np.max(heatmap_0)
print(heatmap_0.shape)

plt.imshow(heatmap_0)
plt.show()
```

* **_Grad-CAM for data1_**
```python
import cv2

DATA1_CLASS_INDEX = 110
test_data_exmpl_1 = test_data[DATA1_CLASS_INDEX,:,:]

test_data_exmpl_1 = np.expand_dims(test_data_exmpl_1, axis=0)
print(test_data_exmpl_1.shape)

conv_layer = model.layers[8]
img_tensor_1 = test_data_exmpl_1
heatmap_model = tf.keras.models.Model([model.inputs], [conv_layer.output, model.output])

# Get the score for target class
with tf.GradientTape() as tape:
    conv_outputs_1, predictions_1 = heatmap_model(img_tensor_1)
    loss_1 = predictions_1[:, np.argmax(predictions_1[0])]
    
# Extract filters and gradients
output_1 = conv_outputs_1[0]
grads_1 = tape.gradient(loss_1, conv_outputs_1)[0]

# Average gradients spatially
weights_1 = tf.reduce_mean(grads_1, axis=(0, 1))
print(weights_1.shape)

# Build a ponderated map of filters according to gradients importance
cam_1 = np.ones(output_1.shape[0:2], dtype=np.float32)
print(cam_1.shape)

for index, w in enumerate(weights_1):
    cam_1 += w * output_1[:, :, index]
    
# Heatmap visualization
cam_1 = cv2.resize(cam_1.numpy(), (128, 120))
cam_1 = np.maximum(cam_1, 0)
heatmap_1 = cam_1
heatmap_1 /= np.max(heatmap_1)
print(heatmap_1.shape)

plt.imshow(heatmap_1)
plt.show()
```

* **_Grad-CAM for data2_**
```python
import cv2

DATA2_CLASS_INDEX = 210
test_data_exmpl_2 = test_data[DATA2_CLASS_INDEX,:,:]

test_data_exmpl_2 = np.expand_dims(test_data_exmpl_2, axis=0)
print(test_data_exmpl_2.shape)

conv_layer = model.layers[8]
img_tensor_2 = test_data_exmpl_2
heatmap_model = tf.keras.models.Model([model.inputs], [conv_layer.output, model.output])

# Get the score for target class
with tf.GradientTape() as tape:
    conv_outputs_2, predictions_2 = heatmap_model(img_tensor_2)
    loss_2 = predictions_2[:, np.argmax(predictions_2[0])]
    
# Extract filters and gradients
output_2 = conv_outputs_2[0]
grads_2 = tape.gradient(loss_2, conv_outputs_2)[0]

# Average gradients spatially
weights_2 = tf.reduce_mean(grads_2, axis=(0, 1))
print(weights_2.shape)

# Build a ponderated map of filters according to gradients importance
cam_2 = np.ones(output_2.shape[0:2], dtype=np.float32)
print(cam_2.shape)

for index, w in enumerate(weights_2):
    cam_2 += w * output_2[:, :, index]
    
# Heatmap visualization
cam_2 = cv2.resize(cam_2.numpy(), (128, 120))
cam_2 = np.maximum(cam_2, 0)
heatmap_2 = cam_2
heatmap_2 /= np.max(heatmap_2)
print(heatmap_2.shape)

plt.imshow(heatmap_2)
plt.show()
```

* **_Grad-CAM for data3_**
```python
import cv2

DATA3_CLASS_INDEX = 310
test_data_exmpl_3 = test_data[DATA3_CLASS_INDEX,:,:]

test_data_exmpl_3 = np.expand_dims(test_data_exmpl_3, axis=0)
print(test_data_exmpl_3.shape)

conv_layer = model.layers[8]
img_tensor_3 = test_data_exmpl_3
heatmap_model = tf.keras.models.Model([model.inputs], [conv_layer.output, model.output])

# Get the score for target class
with tf.GradientTape() as tape:
    conv_outputs_3, predictions_3 = heatmap_model(img_tensor_3)
    loss_3 = predictions_3[:, np.argmax(predictions_3[0])]
    
# Extract filters and gradients
output_3 = conv_outputs_3[0]
grads_3 = tape.gradient(loss_3, conv_outputs_3)[0]

# Average gradients spatially
weights_3 = tf.reduce_mean(grads_3, axis=(0, 1))
print(weights_3.shape)

# Build a ponderated map of filters according to gradients importance
cam_3 = np.ones(output_3.shape[0:2], dtype=np.float32)
print(cam_3.shape)

for index, w in enumerate(weights_3):
    cam_3 += w * output_3[:, :, index]
    
# Heatmap visualization
cam_3 = cv2.resize(cam_3.numpy(), (128, 120))
cam_3 = np.maximum(cam_3, 0)
heatmap_3 = cam_3
heatmap_3 /= np.max(heatmap_3)
print(heatmap_3.shape)

plt.imshow(heatmap_3)
plt.show()
```

### **학습 그래프**🧠

구현한 모델을 통해 학습한 결과 그래프를 accuracy 등과 함께 첨부해주세요.

### 문제점 **🤔**

모델 학습 시 문제점 및 개선 사항이 있다면 적어주세요.
